openshift/dynamic/client.py

```diff
@@ -0,0 +1,387 @@
+            })
+        elif isinstance(field, (list, tuple)):
+            return [self.__deserialize(item) for item in field]
+        else:
+            return field
+
+    def __serialize(self, field):
+        if isinstance(field, ResourceField):
+            return {
+                k: self.__serialize(v) for k, v in field.__dict__.items()
+            }
+        elif isinstance(field, (list, tuple)):
+            return [self.__serialize(item) for item in field]
+        else:
+            return field
+
+    def to_dict(self):
+        return self.__serialize(self.attributes)
+
+    def to_str(self):
+        return repr(self)
+
+    def __repr__(self):
+        return "ResourceInstance[{}]:\n  {}".format(
+            self.attributes.kind,
+            '  '.join(yaml.safe_dump(self.to_dict()).splitlines(True))
+        )
+
+    def __getattr__(self, name):
+        return getattr(self.attributes, name)
+
+    def __getitem__(self, name):
+        return self.attributes[name]
+
+    def __dir__(self):
+        return dir(type(self)) + list(self.attributes.__dict__.keys())
+
+
+class K8sResource(Resource):
+
+    @property
+    def prefix(self):
+        if self._prefix:
+            return self._prefix
+        elif self.apiversion == 'v1' and not self.group:
+            return '/api'
+        else:
+            return '/apis'
+
+
+class OpenshiftResource(Resource):
+
+    @property
+    def prefix(self):
+        if self._prefix:
+            return self._prefix
+        elif self.apiversion == 'v1' and not self.group:
+            return '/oapi'
+        else:
+            return '/apis'
+
+
+def flatten(l):
+    return [item for sublist in l for item in sublist]
+
+
+class DynamicClient(object):
+
+    def __init__(self, client):
+        self.client = client
+        self._groups = self.get_api_groups()
+        self._resources = flatten([self.get_resources_for_group(*group_parts) for group_parts in self._groups])
+
+    def default_groups(self):
+        groups = [('api', '', 'v1', True)]
+
+        try:
+            self.request('get', '/version/openshift')
+            is_openshift = True
+        except ApiException:
+            is_openshift = False
+
+        if is_openshift:
+            groups.append(('oapi', '', 'v1', True))
+        return groups
+
+    def get_api_groups(self):
+        """ Returns a list of API groups in the format:
+            (api_prefix, group_name, version, preferred)
+
+            api_prefix is the url prefix to access the group (usually /apis, but
+            /api for core kubernetes resources and /oapi for core openshift resources)
+
+            group_name and version are the name and version of the group
+
+            preferred is a boolean indicating whether this version of the group is preferred
+        """
+        prefix = 'apis'
+        groups_response = self.request('GET', '/{}'.format(prefix))['groups']
+
+        groups = self.default_groups()
```

Comment:
I'm curious about what is going on here and why we need to detect if the cluster `is_openshift` or not. Would these be missing in the `groups_response`?

Reply:
/apis doesn't return anything that falls under /api or /oapi, so that's why I have the default groups hardcoded.